.circleci/config.yml

```diff
@@ -0,0 +1,140 @@
+    - run:
+        name: "Initialize opam"
+        command: |
+          opam init --auto-setup --dot-profile=~/.bash_profile
+          opam remote add ocamlorg https://opam.ocaml.org -p 0 || true
+          opam remote remove default || true
+    - run:
+        name: "Install deps"
+        command: |
+          opam update
+          opam install -y jbuilder
+          opam install -y menhir.20170712
+    - run:
+        name: 'Clean'
+        command: make clean-for-ci
+    - run:
+        name: 'opam pin'
+        command: |
+          opam pin add -y reason .
+          opam pin add -y rtop .
+    - save_cache:
+        key: dependency-cache-{{ checksum "../ocaml-version" }}-{{ checksum ".circleci/config.yml" }}-{{ checksum "reason.opam" }}-{{ checksum "rebuild.opam" }}-{{ checksum "rtop.opam" }}
+        paths:
+          - ~/.opam
+          - ~/.npm-global
+    - run:
+        name: 'Test'
+        command: |
+          source ~/.bash_profile
+          eval `opam config env`
+          make test-ci
+    - run:
+        name: 'Check git is clean'
+        command: git diff --exit-code
+
+version: 2
+jobs:
+  4.02.3:
+    docker:
+      - image: ocaml/opam:debian-9_ocaml-4.02.3
+    environment:
+      - TERM: dumb
+      - OCAML_VERSION: "4.02.3"
+      - NPM_CONFIG_PREFIX: "~/.npm-global"
+    <<: *common_steps
+  4.03.0:
+    docker:
+      - image: ocaml/opam:debian-9_ocaml-4.03.0
+    environment:
+      - TERM: dumb
+      - OCAML_VERSION: "4.03.0"
+      - NPM_CONFIG_PREFIX: "~/.npm-global"
+    <<: *common_steps
+  4.04.0:
+    docker:
+      - image: ocaml/opam:debian-9_ocaml-4.04.0
+    environment:
+      - TERM: dumb
+      - OCAML_VERSION: "4.04.0"
+      - NPM_CONFIG_PREFIX: "~/.npm-global"
+    <<: *common_steps
+  4.06.0:
+    docker:
+      - image: ocaml/opam:debian-9_ocaml-4.06.0
+    environment:
+      - TERM: dumb
+      - OCAML_VERSION: "4.06.0"
+      - NPM_CONFIG_PREFIX: "~/.npm-global"
+    <<: *common_steps
+  esy_build:
+    docker:
+      - image: ocaml/opam:debian-9_ocaml-4.06.0
+    environment:
+      - TERM: dumb
+      - OCAML_VERSION: "4.06.0"
+      - NPM_CONFIG_PREFIX: "~/.npm-global"
+    steps:
+      - checkout
+      - run:
+          name: "Install npm"
+          command: |
+            curl -sL https://deb.nodesource.com/setup_9.x | sudo -E bash -
+            sudo apt-get install -y nodejs
+            mkdir -p ~/.npm-global
+            npm config set prefix $NPM_CONFIG_PREFIX
+      - run:
+          name: "Note OCAML_VERSION for cache"
+          command: echo $OCAML_VERSION > ../ocaml-version
+      - restore_cache:
+          key: dependency-cache-{{ checksum "../ocaml-version" }}-{{ checksum ".circleci/config.yml" }}-{{ checksum "reason.opam" }}-{{ checksum "rebuild.opam" }}-{{ checksum "rtop.opam" }}
```

Comment:
I think for the `esy_build` portion, all we should need is a checksum on `esy.json`, `esy.lock` and `.circleci/config.yml` file for the cache key. The esy build process has no dependency on opam.
I think you can also remove the `OCAML_VERSION` as well. That's all modeled in the `esy.json` which influences the `esy.lock`. You could probably even move the npm stuff from "common" steps to the esy_build steps since the opam workflow has no dependency on npm/esy.

We can probably continue to use the `ocaml/opam:debian-9_ocaml-4.06.0` image since it will likely be cached and won't harm anything to use it.

Reply:
> You could probably even move the npm stuff from "common" steps to the esy_build steps since the opam workflow has no dependency on npm/esy.

I tried that, but looks like one of the tests requires node. The tests fail at `node ./formatTest/testOprint.js`: https://circleci.com/gh/dwwoelfel/reason/84

Make the recommended updates to the esy_build portion. Thank you!